library/ziti.c

```diff
@@ -2173,21 +2173,29 @@ static void api_session_cb(ziti_api_session *api_sess, const ziti_error *err, vo
             }
         }
 
-        // check if identity cert can and need to be extended
-        if (ztx->opts.cert_extension_window == 0 || ztx->id_creds.cert == NULL) {
+        if (ztx->id_creds.cert == NULL) {
             goto done;
         }
 
-        if (!api_sess->is_cert_extendable) {
-            ZTX_LOG(DEBUG, "identity certificate is not renewable");
-            goto done;
-        }
-
-        struct tm exp;
         if (api_sess->cert_extend_requested || api_sess->key_roll_requested) {
             ZTX_LOG(INFO, "controller requested certificate renewal (%s key roll)",
                     api_sess->key_roll_requested ? "with" : "without");
-        } else {
+            goto extend;
+        }
+
+        if (api_sess->is_cert_improper) {
+            ZTX_LOG(INFO, "controller reported certificate chain as incomplete");
+            goto extend;
+        }
+
+        // check if identity cert is expiring or expired
+        if (ztx->opts.cert_extension_window > 0) {
+            if (!api_sess->is_cert_extendable) {
+                ZTX_LOG(DEBUG, "identity certificate is not renewable");
```

Comment:
seems more like a WARN to me

Reply:
i know you said it before, but it means it is a 3rd party cert, why warn about it?

Reply:
LOL i thought it felt familiar... there's no way to discern if it's from the controller's pki or 3rd party?